# Quality control

After running the download script (versions: [aspera](../LSTrAP-Lite_download.py), [wget](../LSTrAP-Lite_download_wget.py), you should see the following outputs in your directory.
  * <b>logs</b> contains the logfile and error log
  * <b>fastq</b> contains the kallisto outputs
  * <b>matrix_raw.tsv</b> is the gene expression matrix of all downloaded SRA runs
  * <b>mapping_stats.tsv</b> contains informaition regarding the kallisto mapping to help with QC
  * <b>selected_runs.txt</b> empty text file to be updated with the SRA run ID of experiments that you want to continue with after QC.
  
## Mapping statistics
The <b>mapping_stats.tsv</b> file has summarised various information to allow QC of the experiments downloaded. After reviewing the mapping statistics, paste the SRA run IDs of the experiments that you want to proceed with for gene co-expression analysis in <b>selected_runs.txt</b> and save. An example can be found [here](../sample_data/selected_runs.txt).  

## Trimming the gene expression matrix
After modifying <b>selected_runs.txt</b>, run the script [LSTrAP_Lite_TrimMatrix.py](../LSTrAP_Lite_TrimMatrix.py) in the directory defined as "pathtolocal" (refer to the [previous tutorial](01_start.md) if unclear).

<code>python3 LSTrAP_Lite_TrimMatrix.py</code>

You are now ready to move on to creating gene co-expression networks in the [next tutorial](03_genecoexp.md).
